File: Days.32/C++/4.Stock-Span-Problem.cpp
// Stock Span Problem 

#include<iostream>
#include<vector>
#include<stack>
using namespace std;


vector<int> StockSpan(vector<int> &prices,int &n)
{
    vector<int> days(n,1);
    stack<int> stack;
    for (int i=0; i<n; i++)
    {
        while (!stack.empty() and prices[i]>=prices[stack.top()])
        {
            stack.pop();
        }
        if(stack.empty())
        {
            days[i]=i+1;
        }
        else
        {
            days[i]=i-stack.top();
        }
        stack.push(i);
    }
    return days;
}


int main()
{
    int n;
    cin>>n;
    vector<int> prices(n);
    for (int i=0; i<n; i++)
    {
        cin>>prices[i];
    }
    vector<int> days=StockSpan(prices,n);
    for(int i=0; i<n; i++)
    {
        cout<<days[i]<<" ";
    }
    cout<<"\n";
    return 0;
}

File: Days.32/Python/4.Stock-Span-Problem.py
# Stock Span Problem


def StockSpan(prices,n):
    days=[1 for i in range(n)]
    stack=[]
    for i in range(0,n):
        while len(stack)>0 and prices[i]>=prices[stack[-1]]:
            stack.pop()
        if len(stack)==0:
            days[i]=i+1
        else:
            days[i]=i-stack[-1]
        stack.append(i)
    return days


def main():
    n=int(input())
    prices=list(map(int,input().split()))
    days=StockSpan(prices,n)
    for day in days:
        print(day,end=" ")
    print()

main()

